File: llms/llms.py
import os
from github import Github
from github import Auth
import base64

github_token = os.environ.get("GH_API_KEY")
data_dir = "../sites/upsun/public"

class Llms:
    # function to concat config example from the template builder repo (for now: Theosakamg-PSH/template-builder)
    def get_template_builder_configs(self):
        # Create config-examples dir that will contains all downloaded config examples
        config_file_path = data_dir + "/llms-full.txt"

        print(config_file_path)
        auth = Auth.Token(github_token)
        g = Github(auth=auth)
        repo = g.get_repo("Theosakamg-PSH/template-builder")
        try:
            contents = repo.get_contents("templates")
            while contents:
                file_content = contents.pop(0)
                if file_content.type == "dir":
                    # print(file_content.name)
                    template_name = file_content.name
                    print(template_name)
                    # if .upsun folder exists in the remote template directory, add remote config.yaml content into /config-examples/config-examples.md file
                    try:
                        upsun_config_dir = repo.get_contents(file_content.path + '/files/.upsun')
                        while upsun_config_dir:
                            config_file = upsun_config_dir.pop(0)
                            file_data = base64.b64decode(config_file.content).decode('utf-8')
                            config_file_out = open(config_file_path, "a")
                            config_file_out.write('\n## Example of a ' + template_name.capitalize() + ' config \n')
                            config_file_out.write('This is an example of a config.yaml file to host a ``' + template_name + '`` stack on Upsun.\n')
                            config_file_out.write('```yaml {location="config.yaml"}\n')
                            config_file_out.write(file_data)
                            config_file_out.write('\n```\n')
                            config_file_out.close()
                    except Exception as e:
                        print("Error for stack " + template_name + ":", e)
                        pass
        except Exception as e:
            # print("Error:", e)
            pass

if __name__ == "__main__":
    llms = Llms()
    llms.get_template_builder_configs()

File: llms/llms.go
package main

import (
	"context"
	"encoding/base64"
	"fmt"
	"os"
	"strings"

	"github.com/google/go-github/v53/github"
	"golang.org/x/oauth2"
)

const dataDir = "../sites/upsun/public"

// Llms struct
type Llms struct{}

// GetTemplateBuilderConfigs retrieves and processes configuration examples from GitHub
func (l *Llms) GetTemplateBuilderConfigs() {
	configFilePath := dataDir + "/llms-full.txt"
	fmt.Println(configFilePath)

	githubToken := os.Getenv("GH_API_KEY")
	if githubToken == "" {
		fmt.Println("GitHub token not found in environment variables")
		return
	}

	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: githubToken})
	ctx := context.Background()
	client := github.NewClient(oauth2.NewClient(ctx, ts))

	repoOwner := "Theosakamg-PSH"
	repoName := "template-builder"
	dirPath := "templates"
	contents, _, _, err := client.Repositories.GetContents(ctx, repoOwner, repoName, dirPath, nil)
	if err != nil {
		fmt.Println("Error fetching repository contents:", err)
		return
	}

	for _, fileContent := range contents {
		if fileContent.GetType() == "dir" {
			templateName := fileContent.GetName()
			fmt.Println(templateName)

			upsunConfigPath := fileContent.GetPath() + "/files/.upsun"
			upsunContents, _, _, err := client.Repositories.GetContents(ctx, repoOwner, repoName, upsunConfigPath, nil)
			if err != nil {
				fmt.Println("Error for stack", templateName, ":", err)
				continue
			}

			for _, configFile := range upsunContents {
				if configFile.GetType() == "file" {
					content, err := configFile.GetContent()
					if err != nil {
						fmt.Println("Error reading file content:", err)
						continue
					}

					decodedContent, err := base64.StdEncoding.DecodeString(content)
					if err != nil {
						fmt.Println("Error decoding file content:", err)
						continue
					}

					fileData := string(decodedContent)
					output := fmt.Sprintf("\n## Example of a %s config \n", strings.Title(templateName)) +
						"This is an example of a config.yaml file to host a `" + templateName + "` stack on Upsun.\n" +
						"```yaml {location=\"config.yaml\"}\n" + fileData + "\n``\n"

					if err := appendToFile(configFilePath, output); err != nil {
						fmt.Println("Error writing to file:", err)
					}
				}
			}
		}
	}
}

func appendToFile(filename, text string) error {
	file, err := os.OpenFile(filename, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return err
	}
	defer file.Close()

	_, err = file.WriteString(text)
	return err
}

func main() {
	llms := Llms{}
	llms.GetTemplateBuilderConfigs()
}
